substrate/frame/support/src/traits/messages.rs

```diff
@@ -174,6 +176,55 @@ pub struct BatchFootprint {
 	pub new_pages_count: u32,
 }
 
+/// The resource footprints of continuous subsets of messages.
+///
+/// For a set of messages `xcms[0..n]`, each `footprints[i]` contains the footprint
+/// of the batch `xcms[0..i]`, so as `i` increases `footprints[i]` contains the footprint
+/// of a bigger batch.
+#[derive(Default, RuntimeDebug)]
+pub struct BatchesFootprints {
+	pub first_page_pos: usize,
+	pub footprints: Vec<BatchFootprint>,
+}
+
+impl BatchesFootprints {
+	pub fn push(&mut self, msg: &[u8], new_page: bool) {
```

Comment:
Doc please. What happens when this is larger than the page size? I guess it does not need to do any checking since it assumes that the input data is sane?

Reply:
Done. 

> What happens when this is larger than the page size? I guess it does not need to do any checking since it assumes that the input data is sane?

Yes, we don't do any checking here. We consider that the input data is sane, and also, this struct is just a low level helper struct that doesn't have any knowledge about the page size or the message queue. We just add a new batch footprint corresponding to the provided `msg` and `new_page`.  It's the responsibility of the message queue to construct a valid `BatchesFootprints`.